src/audit/logger.rs

```diff
@@ -0,0 +1,221 @@
+        // with the ingest path, This can fail if the URL is not valid,
+        // when the base URL is not set or the ingest path is not valid
+        let Some(url) = CONFIG.parseable.audit_logger.as_ref() else {
+            return logger;
+        };
+
+        logger.log_endpoint = url
+            .join("/api/v1/ingest")
+            .inspect_err(|err| eprintln!("Couldn't setup audit logger: {err}"))
+            .ok();
+
+        // Created directory for audit logs if it doesn't exist
+        std::fs::create_dir_all(&CONFIG.parseable.audit_log_dir)
+            .expect("Failed to create audit log directory");
+
+        // Figure out the latest and oldest log file in directory
+        let files = std::fs::read_dir(&CONFIG.parseable.audit_log_dir)
+            .expect("Failed to read audit log directory");
+        let (oldest_log_file_id, latest_log_file_id) =
+            files.fold((usize::MAX, 0), |(oldest, latest), r| {
+                let file_name = r.unwrap().file_name();
+                let Ok(file_id) = file_name
+                    .to_str()
+                    .expect("File name is not utf8")
+                    .split('.')
+                    .next()
+                    .expect("File name is not valid")
+                    .parse::<usize>()
+                    .inspect_err(|e| warn!("Unexpected file in logs directory: {e}"))
+                else {
+                    return (oldest, latest);
+                };
+                (oldest.min(file_id), latest.max(file_id))
+            });
+
+        logger.next_log_file_id = latest_log_file_id + 1;
+        if oldest_log_file_id != usize::MAX {
+            logger.oldest_log_file_id = oldest_log_file_id;
+        }
+
+        logger
+    }
+}
+
+impl AuditLogger {
+    /// Flushes audit logs to the remote logging system
+    async fn flush(&mut self) {
+        if self.batch.is_empty() {
+            return;
+        }
+
+        // swap the old batch with a new empty one
+        let mut logs_to_send = Vec::with_capacity(CONFIG.parseable.audit_batch_size);
+        std::mem::swap(&mut self.batch, &mut logs_to_send);
+
+        // send the logs to the remote logging system, if no backlog, else write to disk
+        if self.oldest_log_file_id >= self.next_log_file_id
+            && self.send_logs_to_remote(&logs_to_send).await.is_ok()
+        {
+            return;
+        }
+
+        // write the logs to the next log file
+        let log_file_path = CONFIG
+            .parseable
+            .audit_log_dir
+            .join(format!("{}.json", self.next_log_file_id));
+        let mut log_file = OpenOptions::new()
+            .create(true)
+            .truncate(true)
+            .open(log_file_path)
+            .await
+            .expect("Failed to open audit log file");
+        let buf = serde_json::to_vec(&logs_to_send).expect("Failed to serialize audit logs");
+        log_file.write_all(&buf).await.unwrap();
```

Comment:
same

Reply:
This doubt makes sense, will add changes to handle this scenario